File: pkg/compare/compare_test.go
package compare

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/pawel/gha-pin-diff/pkg/diffparser"
	"github.com/pawel/gha-pin-diff/pkg/github"
)

func TestFetch(t *testing.T) {
	mux := http.NewServeMux()
	mux.HandleFunc("/repos/actions/checkout/compare/", func(w http.ResponseWriter, r *http.Request) {
		resp := map[string]interface{}{
			"html_url":      "https://github.com/actions/checkout/compare/aaa...bbb",
			"total_commits": 2,
			"commits": []map[string]interface{}{
				{
					"sha": "abc1234567890abc1234567890abc1234567890ab",
					"commit": map[string]interface{}{
						"message": "Fix something\n\nDetails here",
						"author":  map[string]string{"name": "Alice", "date": "2024-04-15T10:00:00Z"},
					},
					"author": map[string]string{"login": "alice"},
				},
				{
					"sha": "def1234567890def1234567890def1234567890de",
					"commit": map[string]interface{}{
						"message": "Update deps",
						"author":  map[string]string{"name": "Bob", "date": "2024-04-14T09:00:00Z"},
					},
					"author": map[string]string{"login": "bob"},
				},
			},
		}
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(resp)
	})

	srv := httptest.NewServer(mux)
	defer srv.Close()

	client := github.NewClient("")
	client.SetBaseURL(srv.URL)

	updates := []diffparser.ActionUpdate{
		{
			Action: "actions/checkout",
			OldRef: "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
			NewRef: "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb",
			OldTag: "v4.0.0",
			NewTag: "v4.1.0",
			File:   ".github/workflows/ci.yml",
		},
	}

	results := Fetch(context.Background(), client, updates)
	if len(results) != 1 {
		t.Fatalf("expected 1 result, got %d", len(results))
	}

	r := results[0]
	if r.Err != nil {
		t.Fatalf("unexpected error: %v", r.Err)
	}
	if r.TotalCommits != 2 {
		t.Errorf("TotalCommits = %d, want 2", r.TotalCommits)
	}
	if r.CompareURL != "https://github.com/actions/checkout/compare/aaa...bbb" {
		t.Errorf("CompareURL = %q", r.CompareURL)
	}
	if len(r.Commits) != 2 {
		t.Fatalf("len(Commits) = %d, want 2", len(r.Commits))
	}
	if r.Commits[0].Message != "Fix something" {
		t.Errorf("Commits[0].Message = %q, want first line only", r.Commits[0].Message)
	}
	if r.Commits[0].Author != "alice" {
		t.Errorf("Commits[0].Author = %q, want alice", r.Commits[0].Author)
	}
}

func TestFetchAPIError(t *testing.T) {
	mux := http.NewServeMux()
	mux.HandleFunc("/", func(w http.ResponseWriter, _ *http.Request) {
		w.WriteHeader(http.StatusNotFound)
		w.Write([]byte(`{"message":"Not Found"}`))
	})

	srv := httptest.NewServer(mux)
	defer srv.Close()

	client := github.NewClient("")
	client.SetBaseURL(srv.URL)

	updates := []diffparser.ActionUpdate{
		{
			Action: "actions/deleted-action",
			OldRef: "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
			NewRef: "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb",
			File:   ".github/workflows/ci.yml",
		},
	}

	results := Fetch(context.Background(), client, updates)
	if len(results) != 1 {
		t.Fatalf("expected 1 result, got %d", len(results))
	}
	if results[0].Err == nil {
		t.Fatal("expected error for 404 response")
	}
}

func TestActionOwnerRepo(t *testing.T) {
	tests := []struct {
		action    string
		wantOwner string
		wantRepo  string
	}{
		{"actions/checkout", "actions", "checkout"},
		{"org/repo/.github/workflows/deploy.yml", "org", "repo"},
		{"invalid", "", ""},
	}
	for _, tt := range tests {
		owner, repo := actionOwnerRepo(tt.action)
		if owner != tt.wantOwner || repo != tt.wantRepo {
			t.Errorf("actionOwnerRepo(%q) = (%q, %q), want (%q, %q)",
				tt.action, owner, repo, tt.wantOwner, tt.wantRepo)
		}
	}
}

File: src/compare/compare.test.ts
import { describe, expect, it } from "vitest";
import { createServer } from "node:http";
import { fetch as compareFetch, actionOwnerRepo } from "./compare.js";
import { Client, APIError } from "../github/client.js";
import type { ActionUpdate } from "../diffparser/parser.js";

function startServer(
  handler: (req: any, res: any) => void
): Promise<{ url: string; close: () => void }> {
  return new Promise((resolve) => {
    const server = createServer(handler);
    server.listen(0, "127.0.0.1", () => {
      const addr = server.address() as { port: number };
      resolve({
        url: `http://127.0.0.1:${addr.port}`,
        close: () => server.close(),
      });
    });
  });
}

describe("fetch", () => {
  it("fetches comparison data", async () => {
    const { url, close } = await startServer((req, res) => {
      if (req.url?.startsWith("/repos/actions/checkout/compare/")) {
        res.writeHead(200, { "Content-Type": "application/json" });
        res.end(
          JSON.stringify({
            html_url:
              "https://github.com/actions/checkout/compare/aaa...bbb",
            total_commits: 2,
            commits: [
              {
                sha: "abc1234567890abc1234567890abc1234567890ab",
                commit: {
                  message: "Fix something\n\nDetails here",
                  author: { name: "Alice", date: "2024-04-15T10:00:00Z" },
                },
                author: { login: "alice" },
              },
              {
                sha: "def1234567890def1234567890def1234567890de",
                commit: {
                  message: "Update deps",
                  author: { name: "Bob", date: "2024-04-14T09:00:00Z" },
                },
                author: { login: "bob" },
              },
            ],
          })
        );
        return;
      }
      res.writeHead(404);
      res.end();
    });

    try {
      const client = new Client("");
      client.setBaseURL(url);

      const updates: ActionUpdate[] = [
        {
          action: "actions/checkout",
          oldRef: "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
          newRef: "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb",
          oldTag: "v4.0.0",
          newTag: "v4.1.0",
          file: ".github/workflows/ci.yml",
        },
      ];

      const results = await compareFetch(client, updates);
      expect(results).toHaveLength(1);

      const r = results[0];
      expect(r.err).toBeNull();
      expect(r.totalCommits).toBe(2);
      expect(r.compareURL).toBe(
        "https://github.com/actions/checkout/compare/aaa...bbb"
      );
      expect(r.commits).toHaveLength(2);
      expect(r.commits[0].message).toBe("Fix something");
      expect(r.commits[0].author).toBe("alice");
    } finally {
      close();
    }
  });

  it("handles API errors", async () => {
    const { url, close } = await startServer((_req, res) => {
      res.writeHead(404, { "Content-Type": "application/json" });
      res.end(JSON.stringify({ message: "Not Found" }));
    });

    try {
      const client = new Client("");
      client.setBaseURL(url);

      const updates: ActionUpdate[] = [
        {
          action: "actions/deleted-action",
          oldRef: "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
          newRef: "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb",
          oldTag: "",
          newTag: "",
          file: ".github/workflows/ci.yml",
        },
      ];

      const results = await compareFetch(client, updates);
      expect(results).toHaveLength(1);
      expect(results[0].err).not.toBeNull();
      expect(results[0].err).toBeInstanceOf(APIError);
      expect((results[0].err as APIError).statusCode).toBe(404);
    } finally {
      close();
    }
  });
});

describe("actionOwnerRepo", () => {
  it("parses simple action", () => {
    expect(actionOwnerRepo("actions/checkout")).toEqual(["actions", "checkout"]);
  });

  it("parses reusable workflow", () => {
    expect(
      actionOwnerRepo("org/repo/.github/workflows/deploy.yml")
    ).toEqual(["org", "repo"]);
  });

  it("returns null for invalid input", () => {
    expect(actionOwnerRepo("invalid")).toBeNull();
  });
});
